src/virt.rs

```diff
@@ -508,15 +506,27 @@ impl VirtContext {
                         instr
                     );
                     self.handle_device_access_fault(&instr, &device);
+                } else if (self.csr.mstatus & mstatus::MPRV_FILTER) >> mstatus::MPRV_OFFSET == 1 {
+                    // Can an access to device be legally requested with a virtual address?
+                    let instr = unsafe { Arch::get_raw_faulting_instr(&self.trap_info) };
+                    let instr = decode(instr);
+                    let instr_clone = instr.clone();
+                    log::trace!("Access fault with a virtual address: {:x?}", instr);
+                    unsafe {
+                        Arch::handle_virtual_load_store(instr_clone, self);
+                    }
                 } else {
                     log::trace!(
-                        "No matching devices found for address: {:x}",
-                        self.csr.mtval
+                        "No matching device found for address: {:x}",
+                        self.trap_info.mtval
                     );
                     self.emulate_jump_trap_handler();
                 }
             }
             MCause::InstrAccessFault => {
+                let instr = unsafe { Arch::get_raw_faulting_instr(&self.trap_info) };
+                let instr = decode(instr);
+                log::trace!("Instruction access fault: {:x?}", instr);
```

Comment:
Well, currently decoded instruction is required for emulation purposes, is there is a workaround for that?
A bit uncertain about what's bothering you exactly (and why is this use case different from the others), but from my understanding `trap_info.mepc` should be filled by hardware, thus valid, and the question of access address validity kind of falls in the same bucket as miralis/devices protection

Reply:
Yes it is needed when we emulate the instruction, but here this is inside a match arm where we do not emulate an instruction but simply inject a trap. So the `instr` variable is only used for printing in the above snippet.

The problem here is that we have no guarantee that the address holds a valid instruction, because the trap was an instruction access fault, not a trap due to a privileged instruction (which would imply that the instruction is valid).

Reply:
Oh, sorry, my bad, didn't check the lines. In this case yes, you're right.